### PR TITLE
PDTP Privacy metric

### PR DESCRIPTION
**Pull request by abigailgold**
# Description

Added a new privacy metric called PDTP that measures the amount of membership leakage from a model for a set of training samples.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Added new unit tests for this metric.

**Test Configuration**:
- OS: MacOS 10.15.7
- Python version: 3.7
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/classification/classifier.py [hunk 1]
```diff
@@ -136,23 +136,64 @@ class Classifier(ClassifierMixin, BaseEstimator, ABC):
...
+        :return: new estimator
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """
+        Resets the weights of the estimator so that it can be refit from scratch.
+
+        """
+        raise NotImplementedError
```
Comment (beat-buesser) [inline art/estimators/classification/classifier.py:173]: I think these lines should not be necessary.

File: art/estimators/classification/classifier.py [hunk 2]
```diff
@@ -136,23 +136,64 @@ class Classifier(ClassifierMixin, BaseEstimator, ABC):
...
```
Comment (beat-buesser) [inline art/estimators/classification/classifier.py:196]: I think these lines should not be necessary.

File: art/estimators/classification/classifier.py [hunk 3]
```diff
@@ -174,10 +215,38 @@ def save(self, filename: str, path: Optional[str] = None) -> None:
...
+
```
Comment (beat-buesser) [inline art/estimators/classification/classifier.py:232]: I think these lines should not be necessary.

File: art/estimators/classification/classifier.py [hunk 4]
```diff
@@ -174,10 +215,38 @@ def save(self, filename: str, path: Optional[str] = None) -> None:
...
+        :return: new estimator
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """
+        Resets the weights of the estimator so that it can be refit from scratch.
+
+        """
+        raise NotImplementedError
```
Comment (beat-buesser) [inline art/estimators/classification/classifier.py:252]: I think these lines should not be necessary.

File: art/metrics/privacy/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+"""
+Module providing metrics and verifications.
+"""
+from art.metrics.privacy.membership_leakage_metrics import PDTP
```
Comment (beat-buesser) [inline art/metrics/privacy/__init__.py:4]: How about if we change the module name from `membership_leakage_metrics` to `membership_leakage`? I`art.metrics.privacy.membership_leakage` would be shorter and only contain `metrics` once.

File: art/estimators/pytorch.py [hunk 6]
```diff
@@ -114,6 +114,21 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """
+        Resets the weights of the estimator so that it can be refit from scratch.
+
+        """
+        raise NotImplementedError
+
```
Comment (beat-buesser) [inline art/estimators/pytorch.py:131]: I think we should not add these lines as they remove the requirement from classes inheriting from this class of implementing these new methods.

File: art/estimators/scikitlearn.py [hunk 7]
```diff
@@ -26,3 +26,18 @@ class ScikitlearnEstimator(BaseEstimator):
...
+        :return: new estimator
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """
+        Resets the weights of the estimator so that it can be refit from scratch.
+
+        """
+        raise NotImplementedError
```
Comment (beat-buesser) [inline art/estimators/scikitlearn.py:43]: I think we should not add these lines as they remove the requirement from classes inheriting from this class of implementing these new methods.

File: art/estimators/tensorflow.py [hunk 8]
```diff
@@ -73,6 +73,22 @@ def fit(self, x: np.ndarray, y, batch_size: int = 128, nb_epochs: int = 20, **kw
...
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """
+        Resets the weights of the estimator so that it can be refit from scratch.
+
+        """
+        raise NotImplementedError
+
```
Comment (beat-buesser) [inline art/estimators/tensorflow.py:91]: I think we should not add these lines as they remove the requirement from classes inheriting from this class of implementing these new methods.

File: art/estimators/tensorflow.py [hunk 9]
```diff
@@ -160,6 +176,22 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
```
Comment (beat-buesser) [inline art/estimators/tensorflow.py:194]: I think we should not add these lines as they remove the requirement from classes inheriting from this class of implementing these new methods.

File: art/estimators/mxnet.py [hunk 10]
```diff
@@ -81,3 +81,18 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+        :return: new estimator
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """
+        Resets the weights of the estimator so that it can be refit from scratch.
+
+        """
+        raise NotImplementedError
```
Comment (beat-buesser) [inline art/estimators/mxnet.py:98]: I think we should not add these lines as they remove the requirement from classes inheriting from this class of implementing these new methods.

File: art/estimators/keras.py [hunk 11]
```diff
@@ -81,3 +81,18 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+        :return: new estimator
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """
+        Resets the weights of the estimator so that it can be refit from scratch.
+
+        """
+        raise NotImplementedError
```
Comment (beat-buesser) [inline art/estimators/keras.py:98]: I think we should not add these lines as they remove the requirement from classes inheriting from this class of implementing these new methods.

File: art/estimators/classification/keras.py [hunk 12]
```diff
@@ -543,6 +543,14 @@ def fit_generator(self, generator: "DataGenerator", nb_epochs: int = 20, **kwarg
                 kwargs["verbose"] = 0
             super().fit_generator(generator, nb_epochs=nb_epochs, **kwargs)
 
+    def clone_for_refitting(self) -> 'KerasClassifier':  # lgtm [py/inheritance/incorrect-overridden-signature]
```
Comment (beat-buesser) [inline art/estimators/classification/keras.py:546]: I think we should also implement method `reset` here.

File: tests/metrics/privacy/test_membership_leakage_metrics.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+def test_membership_leakage_decision_tree(art_warning, decision_tree_estimator, get_iris_dataset):
+    try:
+        classifier = decision_tree_estimator()
+        extra_classifier = decision_tree_estimator()
+        (x_train, y_train), _ = get_iris_dataset
+        prev = classifier.model.tree_
+        leakage = PDTP(classifier, extra_classifier, x_train, y_train)
+        print(leakage)
+        print(np.average(leakage))
+        print(np.max(leakage))
```
Comment (beat-buesser) [inline tests/metrics/privacy/test_membership_leakage_metrics.py:41]: Would it be possible to log these messages and with `logger`? Similar in the other tests.

File: tests/metrics/privacy/test_membership_leakage_metrics.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+
+@pytest.mark.skipMlFramework("dl_frameworks")
+def test_membership_leakage_decision_tree(art_warning, decision_tree_estimator, get_iris_dataset):
+    try:
+        classifier = decision_tree_estimator()
+        extra_classifier = decision_tree_estimator()
+        (x_train, y_train), _ = get_iris_dataset
+        prev = classifier.model.tree_
+        leakage = PDTP(classifier, extra_classifier, x_train, y_train)
```
Comment (beat-buesser) [inline tests/metrics/privacy/test_membership_leakage_metrics.py:38]: Could it be useful to add an assertion for the expectation on `leakage`?

File: art/metrics/privacy/membership_leakage_metrics.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (beat-buesser) [inline art/metrics/privacy/membership_leakage_metrics.py:1]: Could you please add the license text?

File: art/metrics/privacy/membership_leakage_metrics.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+from typing import TYPE_CHECKING, Optional
+
```
Comment (beat-buesser) [inline art/metrics/privacy/membership_leakage_metrics.py:3]: Could you please add a module docstring?

File: art/metrics/privacy/membership_leakage_metrics.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+if TYPE_CHECKING:
+    from art.estimators.classification import Classifier
+
+
+def PDTP(target_estimator: "Classifier", extra_estimator: "Classifier", x: np.ndarray, y: np.ndarray,
+         indexes: Optional[np.ndarray] = None, num_iter: Optional[int] = 10) -> np.ndarray:
+    """
+        Compute the pointwise differential training privacy metric for the given classifier and training set.
+        Taken from: https://arxiv.org/abs/1712.09136
```
Comment (beat-buesser) [inline art/metrics/privacy/membership_leakage_metrics.py:20]: To follow the pattern in the other docstrings:
```suggestion
        | Paper link: https://arxiv.org/abs/1712.09136
```

File: art/metrics/privacy/membership_leakage_metrics.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+                    "PDTP metric can only be applied to classifiers that implement the reset method."
+                )
+            extra_estimator.fit(alt_x, alt_y)
+            # get probabilities from new model
+            alt_pred = extra_estimator.predict(x)
+            if not is_probability(alt_pred):
+                alt_pred = scipy.special.softmax(alt_pred, axis=1)
+            # divide into 100 bins and return center of bin
+            pred_bin = (np.floor(pred * 100) / 100).round(decimals=2) + 0.005
+            pred_bin[pred_bin > 1] = 0.995
```
Comment (beat-buesser) [inline art/metrics/privacy/membership_leakage_metrics.py:73]: It looks like these lines are independent of `row`, would it be possible to move these lines to before the for-loop in line 56?

Comment (beat-buesser) [reply art/metrics/privacy/membership_leakage_metrics.py:73]: Could it be helpful to use `numpy.histogram` here?

File: art/metrics/privacy/membership_leakage_metrics.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        :param target_estimator: The classifier to be analyzed.
+        :param extra_estimator: Another classifier of the same type as the target classifier, but not yet fit.
+        :param x: The training data of the classifier.
+        :param y: Target values (class labels) of `x`, one-hot-encoded of shape (nb_samples, nb_classes) or indices of
+                  shape (nb_samples,).
+        :param indexes: the subset of indexes of `x` to compute the PDTP metric on. If not supplied, PDTP will be
+                        computed for all samples in `x`.
+        :param num_iter: the number of iterations of PDTP computation to run for each sample. If not supplied,
+                         defaults to 10. The result is the average across iterations.
+        :return: an array containing the average PDTP value for each sample in the training set.
```
Comment (beat-buesser) [inline art/metrics/privacy/membership_leakage_metrics.py:31]: Could you please add an additional sentence describing PDTP like, if I'm right, a higher value means more leakage?

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @abigailgold Thank you very much for your pull request implementing the first privacy metric in ART! I have a few comments, what do you think?

File: art/metrics/privacy/membership_leakage_metrics.py (hunk 18 above)
Comment (abigailgold) [reply art/metrics/privacy/membership_leakage_metrics.py:73]: I don't think that numpy.histogram can help here because I need to know to which bin each sample belongs, but maybe numpy.digitize would work. I'll check.

File: art/estimators/classification/classifier.py (hunk 4 above)
Comment (abigailgold) [reply art/estimators/classification/classifier.py:252]: Ok, I removed them from ClassifierLossGradients, ClassifierClassLossGradients, ClassifierNeuralNetwork and ClassifierDecisionTree. Should I leave them in Classifier or remove as well?

Comment (beat-buesser) [reply art/estimators/classification/classifier.py]: @abigailgold Yes, because they are defined as abstract methods in `BaseEstimator` all estimators will already have inherited then.

File: art/metrics/privacy/membership_leakage_metrics.py (hunk 18 above)
Comment (beat-buesser) [reply art/metrics/privacy/membership_leakage_metrics.py]: Ok.

File: art/estimators/classification/classifier.py (hunk 4 above)
Comment (abigailgold) [reply art/estimators/classification/classifier.py]: Just to make sure, you're saying I should remove them from the Classifier class as well?
